### PR TITLE
Fix bug where you couldn't add a new site in Frequent Sites Mode

### PR DESCRIPTION
**Pull request by fallaciousreasoning**
Resolves https://github.com/brave/brave-browser/issues/24370

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [x] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [x] New files have MPL-2.0 license header
- [x] Adequate test coverage exists to prevent regressions
- [x] Major classes, functions and non-trivial code blocks are well-commented
- [x] Changes in component dependencies are properly reflected in `gn`
- [x] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: browser/ui/webui/new_tab_page/top_sites_message_handler.cc [hunk 1]
```diff
@@ -356,8 +349,12 @@ void TopSitesMessageHandler::HandleAddNewTopSite(
   if (!GetValidURLStringForTopSite(&url))
     return;
 
-  // similar to `MostVisitedHandler::AddMostVisitedTile`
-  if (most_visited_sites_->IsCustomLinksEnabled()) {
-    most_visited_sites_->AddCustomLink(GURL(url), base::UTF8ToUTF16(title));
```
Comment (sangwoo108) [inline browser/ui/webui/new_tab_page/top_sites_message_handler.cc:361]: So, this seems to be the cause! Didn't add custom link when it's in MV mode.

File: browser/ui/webui/new_tab_page/top_sites_message_handler.cc [hunk 2]
```diff
@@ -313,27 +313,20 @@ void TopSitesMessageHandler::HandleEditTopSite(const base::Value::List& args) {
...
+  }
+
   GURL gurl(url);
   GURL new_gurl(new_url);
   std::u16string title16 = base::UTF8ToUTF16(title);
 
-  if (most_visited_sites_->IsCustomLinksEnabled()) {
-    // similar to `MostVisitedHandler::UpdateMostVisitedTile`
-    most_visited_sites_->UpdateCustomLink(
-        gurl, new_gurl != gurl ? new_gurl : GURL(), title16);
```
Comment (sangwoo108) [inline browser/ui/webui/new_tab_page/top_sites_message_handler.cc:323]: cool! I think here was another bug.

File: browser/ui/webui/new_tab_page/top_sites_message_handler.cc [hunk 3]
```diff
@@ -313,27 +313,20 @@ void TopSitesMessageHandler::HandleEditTopSite(const base::Value::List& args) {
...
-
-    // When user tries to edit from frecency mode, try to edit an existing
-    // site, if one exists. If none exists, add a new custom link for the site.
-    bool updated = most_visited_sites_->UpdateCustomLink(
-        gurl, new_gurl, base::UTF8ToUTF16(title));
-    if (!updated) {
-      most_visited_sites_->AddCustomLink(new_url.empty() ? gurl : new_gurl,
-                                         title16);
-    }
+  bool updated = most_visited_sites_->UpdateCustomLink(gurl, new_gurl, title16);
```
Comment (sangwoo108) [inline browser/ui/webui/new_tab_page/top_sites_message_handler.cc:326]: super nit: 
```suggestion
  const bool updated = most_visited_sites_->UpdateCustomLink(gurl, new_gurl, title16);
```

**Review by sangwoo108 — APPROVED** [review]
LGTM 👍

File: browser/ui/webui/new_tab_page/top_sites_message_handler.cc (hunk 1 above)
Comment (fallaciousreasoning) [reply browser/ui/webui/new_tab_page/top_sites_message_handler.cc:361]: Yeah, it was meant to convert to `CustomLinks` mode and then add it but it wasn't :smile: I was talking to rebron, and hopefully we're going to have a bigger rethink of how the new tab page grid is going to work, which should be exciting!

File: browser/ui/webui/new_tab_page/top_sites_message_handler.cc (hunk 3 above)
Comment (fallaciousreasoning) [reply browser/ui/webui/new_tab_page/top_sites_message_handler.cc]: Cool, thanks for the C++ tip! (mine is decidedly below averaged :laughing:)

**Pull request merged**